flink-core/src/main/java/org/apache/flink/api/connector/source/lib/util/IteratorSourceReader.java

```diff
@@ -52,128 +40,19 @@
-    private final SourceReaderContext context;
-
-    /** The availability future. This reader is available as soon as a split is assigned. */
-    private CompletableFuture<Void> availability;
-
-    /**
-     * The iterator producing data. Non-null after a split has been assigned. This field is null or
-     * non-null always together with the {@link #currentSplit} field.
-     */
-    @Nullable private IterT iterator;
-
-    /**
-     * The split whose data we return. Non-null after a split has been assigned. This field is null
-     * or non-null always together with the {@link #iterator} field.
-     */
-    @Nullable private SplitT currentSplit;
-
-    /** The remaining splits that were assigned but not yet processed. */
-    private final Queue<SplitT> remainingSplits;
-
-    private boolean noMoreSplits;
+        extends IteratorSourceReaderBase<E, E, IterT, SplitT> {
 
     public IteratorSourceReader(SourceReaderContext context) {
-        this.context = checkNotNull(context);
-        this.availability = new CompletableFuture<>();
-        this.remainingSplits = new ArrayDeque<>();
-    }
-
-    // ------------------------------------------------------------------------
-
-    @Override
-    public void start() {
-        // request a split if we don't have one
-        if (remainingSplits.isEmpty()) {
-            context.sendSplitRequest();
-        }
-    }
-
-    @Override
-    public InputStatus pollNext(ReaderOutput<E> output) {
-        if (iterator != null) {
-            if (iterator.hasNext()) {
-                output.collect(iterator.next());
-                return InputStatus.MORE_AVAILABLE;
-            } else {
-                finishSplit();
-            }
-        }
-
-        return tryMoveToNextSplit();
-    }
-
-    private void finishSplit() {
-        iterator = null;
-        currentSplit = null;
-
-        // request another split if no other is left
-        // we do this only here in the finishSplit part to avoid requesting a split
-        // whenever the reader is polled and doesn't currently have a split
-        if (remainingSplits.isEmpty() && !noMoreSplits) {
-            context.sendSplitRequest();
-        }
-    }
-
-    private InputStatus tryMoveToNextSplit() {
-        currentSplit = remainingSplits.poll();
-        if (currentSplit != null) {
-            iterator = currentSplit.getIterator();
-            return InputStatus.MORE_AVAILABLE;
-        } else if (noMoreSplits) {
-            return InputStatus.END_OF_INPUT;
-        } else {
-            // ensure we are not called in a loop by resetting the availability future
-            if (availability.isDone()) {
-                availability = new CompletableFuture<>();
-            }
-
-            return InputStatus.NOTHING_AVAILABLE;
-        }
+        super(context);
     }
 
     @Override
-    public CompletableFuture<Void> isAvailable() {
-        return availability;
-    }
+    protected void start(SourceReaderContext context) {}
```

Comment:
Should this be really in here?

Reply:
Good question :sweat_smile: 
![image](https://user-images.githubusercontent.com/1492164/195075833-21b06f19-4520-4a05-8d9d-68cfc2744f4f.png)

We can push the no-op `start()` to the base class instead of having it abstract, but that won't work for `convert()` since it deals with the `E->E / E->O` distinction for non-mapping source readers.

Reply:
I was confused; I thought. the IteratorSourceReader was the parent class.

This is fine as is imo.